### PR TITLE
Fix SelectionFunction double backwards by not unwrapping grad_output …

### PR DESCRIPTION
**Pull request by gchanan**
…data.

### CODE REVIEW

**Comment by gchanan** [comment]
Fixes https://github.com/pytorch/pytorch/issues/1712.

I verified this using gradgradcheck from https://github.com/pytorch/pytorch/pull/1643 -- it fails without this change and passes with it.

File: torch/autograd/_functions/reduce.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def forward(cls, ctx, input, dim=None, keepdim=True, additional_args=tuple()):
     def backward(cls, ctx, grad_output, grad_indices=None):
         grad_input = Variable(grad_output.data.new(*ctx.input_size).zero_())
         if ctx.dim is None and cls.has_all_reduce:
-            grad_input[ctx.indices_tuple] = grad_output.data[0]
+            grad_input[ctx.indices_tuple] = grad_output[0]
```
Comment (apaszke) [inline torch/autograd/_functions/reduce.py]: Isn't `[0]` a no-op here?

Comment (gchanan) [inline torch/autograd/_functions/reduce.py]: It ends up calling Index.apply, right?

>>> y=Variable(torch.randn(5,6), requires_grad=True)
>>> y[0].grad_fn
<torch.autograd.function.IndexBackward object at 0x7fa2c45e9de0>

Comment (apaszke) [inline torch/autograd/_functions/reduce.py]: It does, but the Index is completely unnecessary

Comment (gchanan) [inline torch/autograd/_functions/reduce.py]: right!  I clearly took "no-op" too literally...

**Pull request merged**